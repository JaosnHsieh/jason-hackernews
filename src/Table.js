import React from "react";
import Button from "./Button.js";
const Table = ({ list, onDismiss }) => (
  <div>
    {list.map(ele => (
      <div className="list" key={ele.objectID}>
        <Button onClick={() => onDismiss(ele.objectID)}>X</Button>
        <span>{ele.title}</span>
        <span>{ele.author}</span>
        <span>{ele.points}</span>
      </div>
    ))}
  </div>
);

export default Table;
