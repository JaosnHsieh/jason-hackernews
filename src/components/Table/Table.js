import React, { Component } from "react";
import PropTypes from "prop-types";
import { orderBy } from "lodash";
import Sort from "../Sort";

const SORTS = {
  NONE: list => [...list],
  AUTHOR: list => orderBy(list, "author", ["asc"]),
  TITLE: list => orderBy(list, "title", ["asc"]),
  COMMENTS: list => orderBy(list, "num_comments", ["asc"]),
  POINTS: list => orderBy(list, "points", ["asc"])
};

class Table extends Component {
  constructor(props) {
    super(props);
    this.state = {
      sortKey: "NONE",
      isReverseSort: false
    };
  }
  onSort = sortKey => {
    const isReverseSort =
      this.state.sortKey === sortKey && !this.state.isReverseSort;
    this.setState({
      sortKey,
      isReverseSort
    });
  };
  render() {
    const { list, onDismiss } = this.props;

    const sortedList = SORTS[this.state.sortKey](list);
    const reverseSortedList = this.state.isReverseSort
      ? sortedList
      : sortedList.reverse();
    return (
      <div>
        <div className="list">
          <span className="column1" />
          <span className="column2">
            <Sort
              sortKey={"TITLE"}
              onSort={this.onSort}
              activeSortKey={this.state.sortKey}
              isReverseSort={this.state.isReverseSort}
            >
              TITLE
            </Sort>
          </span>
          <span className="column3">
            <Sort
              sortKey={"AUTHOR"}
              onSort={this.onSort}
              activeSortKey={this.state.sortKey}
              isReverseSort={this.state.isReverseSort}
            >
              AUTHOR
            </Sort>
          </span>
          <span className="column4">
            <Sort
              sortKey={"COMMENTS"}
              onSort={this.onSort}
              activeSortKey={this.state.sortKey}
              isReverseSort={this.state.isReverseSort}
            >
              COMMENTS
            </Sort>
          </span>
          <span className="column5">
            <Sort
              sortKey={"POINTS"}
              onSort={this.onSort}
              activeSortKey={this.state.sortKey}
              isReverseSort={this.state.isReverseSort}
            >
              POINTS
            </Sort>
          </span>
        </div>
        {reverseSortedList.map(ele => (
          <div className="list" key={ele.objectID}>
            <span className="column1">
              <a
                className="button-delete"
                onClick={() => onDismiss(ele.objectID)}
              >
                X
              </a>
            </span>
            <span className="column2">
              <a href={ele.url}>{ele.title}</a>
            </span>
            <span className="column3">{ele.author}</span>
            <span className="column4">{ele.num_comments}</span>
            <span className="column5">{ele.points}</span>
          </div>
        ))}
      </div>
    );
  }
}

Table.propTypes = {
  list: PropTypes.arrayOf(
    PropTypes.shape({
      objectID: PropTypes.string.isRequired,
      author: PropTypes.string,
      url: PropTypes.string,
      points: PropTypes.number
    })
  ).isRequired
};
export default Table;
