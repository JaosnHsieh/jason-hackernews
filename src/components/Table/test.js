import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import Table from "./Table";
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

Enzyme.configure({ adapter: new Adapter() });

describe("Table", () => {
  const props = {
    list: [
      {
        objectID: "objid-test1",
        title: "title1",
        author: "jason1",
        points: 100
      },
      {
        objectID: "objid-test2",
        title: "title2",
        author: "jason2",
        points: 101
      },
      {
        objectID: "objid-test3",
        title: "title3",
        author: "jason3",
        points: 102
      }
    ]
  };
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<Table {...props} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it("has a valid snapshot?", () => {
    const component = renderer.create(<Table {...props} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("show 3 items in the list", () => {
    const element = shallow(<Table {...props} />);
    expect(element.find("div.list").length).toBe(props.list.length + 1); // +1 is for Table-header
  });
});
