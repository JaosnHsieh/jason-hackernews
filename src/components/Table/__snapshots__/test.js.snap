// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table has a valid snapshot? 1`] = `
<div>
  <div
    className="list"
  >
    <span
      className="column1"
    >
      Delete
    </span>
    <span
      className="column2"
    >
      <button
        className="sort-button"
        onClick={[Function]}
        type="button"
      >
        TITLE
      </button>
    </span>
    <span
      className="column3"
    >
      <button
        className="sort-button"
        onClick={[Function]}
        type="button"
      >
        AUTHOR
      </button>
    </span>
    <span
      className="column4"
    >
      <button
        className="sort-button"
        onClick={[Function]}
        type="button"
      >
        COMMENTS
      </button>
    </span>
    <span
      className="column5"
    >
      <button
        className="sort-button"
        onClick={[Function]}
        type="button"
      >
        POINTS
      </button>
    </span>
  </div>
  <div
    className="list"
  >
    <span
      className="column1"
    >
      <button
        className=""
        onClick={[Function]}
        type="button"
      >
        X
      </button>
    </span>
    <span
      className="column2"
    >
      title3
    </span>
    <span
      className="column3"
    >
      jason3
    </span>
    <span
      className="column4"
    />
    <span
      className="column5"
    >
      102
    </span>
  </div>
  <div
    className="list"
  >
    <span
      className="column1"
    >
      <button
        className=""
        onClick={[Function]}
        type="button"
      >
        X
      </button>
    </span>
    <span
      className="column2"
    >
      title2
    </span>
    <span
      className="column3"
    >
      jason2
    </span>
    <span
      className="column4"
    />
    <span
      className="column5"
    >
      101
    </span>
  </div>
  <div
    className="list"
  >
    <span
      className="column1"
    >
      <button
        className=""
        onClick={[Function]}
        type="button"
      >
        X
      </button>
    </span>
    <span
      className="column2"
    >
      title1
    </span>
    <span
      className="column3"
    >
      jason1
    </span>
    <span
      className="column4"
    />
    <span
      className="column5"
    >
      100
    </span>
  </div>
</div>
`;
